specs/Standards/AccountRegistration.md

```diff
@@ -0,0 +1,89 @@
+# Account Registration
+
+Version `1.0.0`
+
+## Summary
+
+TODO
+
+## Motivation
+
+To simplify the interface we and make sure we send funds only to accounts which agree to receive token of a given type we introduce account registration concept. NEP-21 transfer (and approval) functions require payment to cover the potential storage fees. 
+
+With _account registration_, an account must firstly to opt-in to a token, and only then it can receive tokens. This gives many benefits:
+
+* protects account for receiving unwanted tokens (spamming, compliance)
+* signals that a smart-contract can handle tokens
+
+NOTE: Each method which calls a #payment method must be marked with #payment. This complicates the tools design - they will need to ask user to attach payment for every such transaction.  However we decided to keep the payment requirement for security reasons related to wallet and transaction signing. Please look at NEP-141 (Fungible Token Core Standard) for details.
+
+An account, once registered, can opt-out - with that, a storage deposit will be returned, and the account will stop accepting token transfers.
+
+## Guide-level explanation
+
+TODO
+
+### Example scenarios
+
+TODO
+
+
+## Reference-level explanation
+
+**NOTES**:
+
+- All amounts, balances and allowance are limited by `U128` (max value 2<sup>128</sup> - 1).
+- This storage standard uses JSON for serialization of arguments and results.
+- Amounts in arguments and results are serialized as Base-10 strings, e.g. `"100"`. This is done to avoid JSON limitation of max integer value of 2<sup>53</sup>.
+- To prevent the deployed contract from being modified or deleted, it should not have any access keys on its account.
+
+**Interface**:
+
+```ts
+/***************************************/
+/* CHANGE METHODS on Account Registrar */
+/***************************************/
+// Registers an account for using smart-contract. Must attach amount of NEAR given by `ar_registration_fee`.
+// Parameters:
+// * `account_id` - an account to register. Anyone can register any other
+//   account.
+// If the account is already registered then the function early returns and
+// refunds the attached NEAR.
```

Comment:
Do we want to handle refund if the attached deposit is larger than the `ar_registration_fee`?

E.g. from current rust implementaiton:
```
 MUST refund the unused attached amount, e.g.
    - if the account already registered, it MUST return the full attached amount.
    - if the attached account is larger than the `ar_registration_fee` should refund extra.
```

Reply:
we addressed this with min/max using bounds